### ISSUE TITLE
If there is an error when fetching a listing, return a 500 level code instead of 404ing

### ISSUE THREAD

**Issue by rmisio** [OpenBazaar/openbazaar-go#514]
Observed:
If fetching a listing that is invalid or an error occurs during the fetch (e.g. too many open files), the API is returning a 404.

Expected:
The API returns 500 with the error that happened.

By convention 404 is used to indicate a resource is not found. The client will show a generic "Listing not found" page in this scenario. If the listing is found, but there's an error, a 500 error code will allow the client to show an error page and display the error. This is much cleaner from a debugging perspective and, IMHO from a UX perspective so the user understands the listing they are seeking is indeed a listing, it just cannot be retrieved at the moment.

**Comment by cpacia** [comment]
I'm honestly not sure what to do with this one. If the result of ipfs.Cat() returns an error then I return NotFound as ipfs couldn't find the file. Anything else is returned as a 500 error. 

It could be that IPFS returns some other error like `too many files open` but to tell the difference between them I need to trace the function calls back like 10-12 layers through the entire IPFS library to find which errors are being passed and when.

Is this really needed functionality to distinguish between fine grained IPFS errors? It seems like NOT FOUND is sufficient to capture errors which result in IPFS not returning the needed data.

**Comment by rmisio** [comment]
> Is this really needed functionality to distinguish between fine grained IPFS errors? 

There's some cases where not being able to distinguish is particularly problematic. One example is fetching our own profile on start-up. If a profile is not set, we will onboard the user. But in a scenario like this where the profile may actually be set but the API is 404ing, the user would needlessly need to go through onboarding again. And... it's possible some other parts of their profile would be overwritten, since onboarding only affects a subset of the profile and we default out the rest.

> but to tell the difference between them I need to trace the function calls back like 10-12 layers through the entire IPFS library to find which errors are being passed and when.

What would the scope for this be? Are we talking 4 hours or 4 days? Maybe we could ask the IPFS team if there is an easier way?

**Comment by cpacia** [comment]
If it's your own profile it doesn't query IPFS, instead it just loads from disk so there is no chance that an IPFS error would be returned.

**Comment by rmisio** [comment]
> If it's your own profile it doesn't query IPFS, instead it just loads from disk so there is no chance that an IPFS error would be returned.

That is not true. I've experienced the scenario I described many times with a 'too many open files' error. In that case a user with an already set Profile will be shuttled to on-boarding again.

To get some perspective, what would be the scope here to be able to distinguish from a resource not being set (a true 404) vs. an IPFS error in retrieving it?
